src/hotspot/cpu/riscv/riscv.ad

```diff
@@ -6598,6 +6598,26 @@ instruct negL_reg(iRegLNoSp dst, iRegL src, immL0 zero) %{
   ins_pipe(ialu_reg);
 %}
 
+// Combined Multiply-Add Shorts into Integer (dst = src1 * src2 + src3 * src4)
+// It's not necessary to have this one, as there is no instructions on riscv to do a
+// scalar muladd. But to enable MulAddVS2VI instrinsic, we need the scalar version first.
+
+instruct muladdS2I(iRegINoSp dst, iRegIorL2I src1, iRegIorL2I src2, iRegIorL2I src3, iRegIorL2I src4) %{
+  match(Set dst (MulAddS2I (Binary src1 src2) (Binary src3 src4)));
+
+  ins_cost(IMUL_COST * 2);
+  format %{ "mulw  rscratch1, $src1, $src2\n\t"
+            "maddw $dst, $src3, $src4, rscratch1" %}
+
+  ins_encode %{
+    __ mul(t0, as_Register($src1$$reg), as_Register($src2$$reg));
+    __ mul(t1, as_Register($src3$$reg), as_Register($src4$$reg));
```

Comment:
Note that it's risky to use `t1` here as it's the emulated flags register for C2 on riscv. So you might want to reserve another temporary register to replace it.

Reply:
You're right.

I'm not quite familiar with this part. Just a question about the flag register (t1) in riscv, do we already use t1 as flag register in any code? I've worked on something related to it, but seems it's not performant (https://bugs.openjdk.org/browse/JDK-8320989)

Reply:
Yes and you will find some usages if you grep "Set cr" in file riscv.ad: 
```
  match(Set cr (CmpP (PartialSubtypeCheck sub super) zero));
  match(Set cr (FastLock object box));
  match(Set cr (FastUnlock object box));
  match(Set cr (FastLock object box));
  match(Set cr (FastUnlock object box));
```

These C2 nodes are expecting a return value in the flag register (aka `rFlagsReg cr`):
```
 3814 // Flags register, used as output of compare logic
 3815 operand rFlagsReg()
 3816 %{
 3817   constraint(ALLOC_IN_RC(reg_flags));
 3818   match(RegFlags);
 3819
 3820   op_cost(0);
 3821   format %{ "RFLAGS" %}
 3822   interface(REG_INTER);
 3823 %}
```

Reply:
Thanks, I've updated it to use a passed in tmp reg.